scripts/find-documents-missing-metadata.ts

```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/node
+
+// This script aims to fix certain documents that were created in Firestore with missing metadata.
+// The sympom is metadata documents that are missing their unit, investigation, and problem fields.
+// In at least some cases, there is another document with the same document key (but different network)
+// that has the correct metadata. This script will identify those documents and copy the metadata
+// to the document where it is missing.
+
+// Set dryRun to false below to actually make changes to the database.
+
+// to run this script type the following in the terminal
+// $ cd scripts
+// $ npx tsx find-documents-missing-metadata.ts
+
+import admin from "firebase-admin";
+import { getFirestoreBasePath, getScriptRootFilePath } from "./lib/script-utils.js";
+
+// The portal to get documents from. For example, "learn.concord.org".
+const portal = "learn.concord.org";
+// The demo name to use. Make falsy to not use a demo.
+// const demo = "CLUE";
+const demo = false;
+
+// Limit number of documents returned from query; or set to false to include all documents
+const documentLimit = 10;
+
+const dryRun = true;
+
+const should_match_fields = [ "context_id", "key", "title", "type" ];
+const fields_to_copy = [ "unit", "investigation", "problem" ];
+
+const databaseURL = "https://collaborative-learning-ec215.firebaseio.com";
+
+// Fetch the service account key JSON file contents
+const credential = admin.credential.cert(getScriptRootFilePath("serviceAccountKey.json"));
+// Initialize the app with a service account, granting admin privileges
+admin.initializeApp({
+  credential,
+  databaseURL
+});
+
+console.log(`*** Starting to Scan Documents ***`);
+
+const collectionUrl = getFirestoreBasePath(portal, demo);
+
+const documentCollection = admin.firestore().collection(collectionUrl);
+
+let documentQuery = documentCollection
+  // Omit document types that are not expected to be associated with a unit, investigation, and problem
+  .where("type", "not-in", ["personal", "personalPublication", "learningLog", "learningLogPublication"])
+  .where("unit", "==", null);
+
+if (documentLimit) {
+  documentQuery = documentQuery.limit(documentLimit);
+}
+
+const singles = new Set();
+const triplets = new Set();
+const pairs = new Set();
+const mismatchDocs = new Set();
+const fixableDocsByType = { "problem": 0, "problemPublication": 0, "planning": 0 };
+
+const promises = [];
+
+const documentSnapshots = await documentQuery.get();
+documentSnapshots.forEach(doc => {
+  const data = doc.data();
+  // console.log(`Document ${doc.ref.path}: ${data.type} unit: ${data.unit}`);
+  // Look for other documents with the same key
+  const key = data.key;
+  const keyQuery = documentCollection.where("key", "==", key);
+  const query = keyQuery.get().then(async snapshot => {
+    if (snapshot.size === 1) {
+      singles.add(key);
+    } else if (snapshot.size > 2) {
+      triplets.add(key);
```

Comment:
More than 2 documents is valid in our current system. But perhaps it really rare.

Reply:
I put in this line to count them if they exist; script did not find any.